test/sslapitest.c

```diff
@@ -4555,6 +4556,157 @@ static int test_ssl_get_shared_ciphers(int tst)
+            || !TEST_int_eq(memcmp(tickdata, appdata, tickdlen), 0))
+        return SSL_TICKET_FATAL_ERR_OTHER;
+
+    return retv;
+}
+
+static int tick_renew = 0;
+
+static int tick_key_cb(SSL *s, unsigned char key_name[16],
+                       unsigned char iv[EVP_MAX_IV_LENGTH], EVP_CIPHER_CTX *ctx,
+                       HMAC_CTX *hctx, int enc)
+{
+    const unsigned char tick_aes_key[16] = "0123456789abcdef";
+    const unsigned char tick_hmac_key[16] = "0123456789abcdef";
+
+    memset(iv, 0, AES_BLOCK_SIZE);
+    memset(key_name, 0, 16);
+    if (!EVP_CipherInit_ex(ctx, EVP_aes_128_cbc(), NULL, tick_aes_key, iv, enc)
+            || !HMAC_Init_ex(hctx, tick_hmac_key, sizeof(tick_hmac_key),
+                             EVP_sha256(), NULL))
+        return -1;
+
+    return tick_renew ? 2 : 1;
+}
+
+/*
+ * Test the various ticket callbacks
+ * Test 0: TLSv1.2, no ticket key callback (default no ticket renewal)
+ * Test 1: TLSv1.3, no ticket key callback (default ticket renewal)
+ * Test 2: TLSv1.2, ticket key callback, no ticket renewal
+ * Test 3: TLSv1.3, ticket key callback, no ticket renewal
+ * Test 4: TLSv1.2, ticket key callback, ticket renewal
+ * Test 5: TLSv1.3, ticket key callback, ticket renewal
+ */
+static int test_ticket_callbacks(int tst)
+{
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    SSL_SESSION *clntsess = NULL;
+    int testresult = 0;
+
+#ifdef OPENSSL_NO_TLS1_2
+    if (tst % 2 == 0);
+        return 1;
+#endif
+#ifdef OPENSSL_NO_TLS1_3
+    if (tst % 2 == 1);
+        return 1;
+#endif
+
+    gen_tick_called = dec_tick_called = 0;
+
+    /* Which tests the ticket key callback should request renewal for */
+    if (tst == 4 || tst == 5)
+        tick_renew = 1;
+    else
+        tick_renew = 0;
+
+    if (!TEST_true(create_ssl_ctx_pair(TLS_server_method(),
+                                       TLS_client_method(),
+                                       TLS1_VERSION,
+                                       tst == 0 ? TLS1_2_VERSION
+                                                : TLS1_3_VERSION,
+                                       &sctx, &cctx, cert, privkey)))
+        goto end;
+
+    if (!TEST_true(SSL_CTX_set_session_ticket_cb(sctx, gen_tick_cb, dec_tick_cb,
+                                                 NULL)))
+        goto end;
+
+    if (tst >= 2
+            && !TEST_true(SSL_CTX_set_tlsext_ticket_key_cb(sctx, tick_key_cb)))
+        goto end;
+
+    if (!TEST_true(create_ssl_objects(sctx, cctx, &serverssl, &clientssl,
+                                             NULL, NULL))
+            || !TEST_true(create_ssl_connection(serverssl, clientssl,
+                                                SSL_ERROR_NONE)))
+        goto end;
+
+    if (!TEST_int_eq(gen_tick_called, 1)
+            || !TEST_int_eq(dec_tick_called, 0))
+        goto end;
+
+    gen_tick_called = dec_tick_called = 0;
+
+    clntsess = SSL_get1_session(clientssl);
```

Comment:
Can we be sure that with TLS 1.3 the server's ticket message has been received by this point?  I'm not sure how far beyond `TLS finished` `create_ssl_objects` proceeds?

Reply:
Yes. `create_ssl_connection` has an explicit client side read at the end of it to make sure we've processed any NewSessionTicket that the server might have sent, so we can rely on the session being available at this point.

Reply:
Perhaps the callback return value space should be separate from the input `retv` space. The callback can't make decryption appear to succeed after it failed (that might even lead to crashes later).  The only possible options as I see them are:

1. Abort the handshake
2. Continue, but don't generate a new ticket when one would have been generated otherwise.
3. Continue, as planned

These could be the &lt;0, 0 and &gt;0 return cases. These could be *combined* with the pending return value to decide what to do next...

Reply:
There is probably also: continue, but always generate a new ticket even if we wouldn't have done otherwise.

Reply:
> @mattcaswell There is probably also: continue, but always generate a new ticket even if we wouldn't have done otherwise.

If so, then we can just let the callback decide to abort, continue with new ticket or continue without new ticket, and it can check the intended course of action from the passed in `retv`.

Note that the earlier code that decrypted the ticket is the one that knows wether the keys were fresh or stale, ... so one really should re-issue the ticket if the key is about to be retired. Otherwise, the application may continue to use the same ticket past is server-side expiration. So perhaps the better option is to be able to force generation of a new ticket, but suppression is not necessarily a good idea.

A more flexible trichotomy is abort, continue without new ticket, continue with new ticket, and the new callback decides based on `retv` and what it finds in any decrypted session (a NULL pointer should be passed for the session when nothing was decrypted, we currently only update *psess on success, probably should set it NULL on failure).

We might also note that there's a difference between not having a matching key (based on key name) and decryption failing even though we should have had the right key, but that's not reflected in values passed to the new callback. Not all values of `SSL_TICKET_NO_DECRYPT` are alike, so the new callback can't tell the difference between stale tickets where we no longer have a matching key and mangled tickets where the key name is a match, but the payload is damaged or bogus.

Reply:
In this case, then, the documentation should have an example callback. I'm concerned that this may be harder to understand by users, than simply passing back the original return value (which implies making the decision). But, I suspect a "hint" would be better, so that no protocol specifications are broken.

Reply:
> @tmshort I'm concerned that this may be harder to understand by users, than simply passing back the original return value (which implies making the decision). But, I suspect a "hint" would be better, so that no protocol specifications are broken.

If the callback should always just return its input value, then it should just be `void`. So, in fact, giving it the freedom to return any of the same range of values Is the *_most_* complex design. Indeed it even makes possible for the appearance of successful decryption when none took place, possibly causing problems later.

There are only three possible outcomes, abort, continue without new ticket, continue with new ticket.  The callback's return value should *not* mirror the decryption status, it can't retroactively make something decrypt successfully that did not.

Please document clearly the actions that the callback can take, and then we can base the implementation on the documented available actions.

Reply:
Do not confuse the return value of the callback with the return value of its caller.  The caller will process the callback's return value and choose its own return value appropriately.  The callback should not know the return-value contract of its caller, and will not be a tall-call.

Reply:
> If the callback should always just return its input value, then it should just be void.

That is absolutely not the case, but can be used by a callback that just wants to get the application data. The intent was to manipulate ticket use based on the application data (and other parameters).

Regardless, it sounds as though this is going to change.

One option that is useful, but not yet mentioned in this discussion, is to *IGNORE* the ticket (i.e. do a full handshake); I don't see that's possible with the trichotomy option (the original code would have used SSL_TICKET_NONE as a return).

Thus, I see at least 5 possible results from the callback:

1. Abort the handshake (-1)
1. Ignore the ticket, do not issue a new ticket (0)
1. Ignore the ticket, issue a new ticket (1)
1. Use the ticket, do not issue a new ticket (2)
1. Use the ticket, issue a new ticket (3)

Reply:
If we do need all 5 outcomes they should still use a separate set of values from the input states. We must not coerce an exact correspondence between the callback's return values and the return-value contract of its caller.  Note also that "use the ticket" makes no sense when there is no ticket to use, so the caller would have treat 4/5 as identical to 2/3 when no ticket was obtained.

Reply:
I'm AOK with that. And yes, if there was no ticket, then "Use the ticket" makes no sense; and should be ignored.